### PR TITLE
Promote weak form SINDy from a feature library to a sibling `SINDy` class.

### PR DESCRIPTION
**Pull request by Jacob-Stevens-Haas**
This PR, long in the making, refactors the weak form.  Previously, weak form was implemented as a feature library, and the LHS calculation was guarded by an "if weak: do this, else: calculate derivatives".  However, the weak form is a lot more than just a feature library, and the implementation forced a lot of "spooky action at a distance" in other classes and functions, which meant that touching those classes could cause weak sindy to break (e.g. #395).  See #351 for more overall motivation.

In order for PDELibrary to support this change, it was easier to remove SINDy-PI from the PDELibrary and make it its own class, as discussed in in #192.  This PR implements the recommended fix.

## What do we get from refactoring WeakSINDy?
* For one, we instantly gain the ability to fit multiple trajectories (with potentially different spatiotemporal grids), since `st_grid` in `WeakSINDy` is analagous to `t` in `SINDy`.
* We also gain the ability to use `predict()` in weak problems, because they use normal feature libraries.  This includes simulate and score for ODEs.
* We can use smoothing from different differentiation methods on the data before forming weak features.
* Special guards in `GeneralizedLibrary` for tensoring the weak form with a regular library are no longer needed.  I believe there were also cases where they may also have been wrong in cases.
* Along the way, I've isolated a lot of the functionality to support unit tests.  These give confidence that I'm keeping things mathematically correct for all of the quadrature/integration by parts.  That rigor also allows us to accept a user-supplied test function or test PRs of alternate test functions.
* Clarifies how to use control features and steady-state solutions (e.g. $$f(x) = Du$$)
* Choosing subdomains is factored into a helper function that returns a `SubdomainSpecs` dataclass.  People can implement alternate strategies (as mentioned in the original Messenger and Bortz paper) for choosing subdomains over the legacy random choice.
* The use of a separate SINDy class allows one to implement the noise calculations of Messenger and Bortz to pass to a `BINDy` like algorithm (See #674)

## What does it cost?
`WeakPDELibrary` is maintained but deprecated because the new `WeakSINDy` class is not compatible with
* SINDy-PI.  To do this correctly requires more explicitly symbolic input feature names, i.e. recognizing that `u_1` is a derivative of the first spatial dimension, whereas `u1` is the first system coordinate.  This PR is a step in that direction, but ultimately we will need to use `sympy`. 
* `GeneralizedLibrary`.  However, this PR is is correct with odd tensored/concat constructions of libraries, due to the flattening/distributive operation, where `GeneralizedLibrary` was not.  But because of the distributive operation, it's unclear how `inputs_per_library` would work.  Perhaps if `inputs_per_library` were passed in `transform()` (after all, `input_features` come in `fit`, not `__init__`) it would be easier.   

## Developer notes
The _weak.py module begins with a docstring describing the relevant helper objects.
* Added performance benchmarks in CI for `SINDy` and `WeakSINDy` runtime, peakmemory, and fit/predict performance.
* Weights are no longer stored in a 2D array, indexed by derivative operation and subdomain.  Instead, they are stored in a dictionary mapping from derivative op to a 1D list of weights on each subdomain.  Removing an indexing convention makes the lookups clearer.
* Throughout the repo, t and st_grid now follow the same convention as most other arrays.  Everything gets promoted into a minimum 2D AxesArray, with the second-to-last axis is time and the last axis is coordinate.
* All tensored/concatenated libraries are distributed so as to make  integration by parts and term ordering clearer.
* `SemiTerm` is an internal object to keep track of an integral term symbolically.  They are `Semi` because sometimes multiple need to be summed up to get a single feature - this was always the case, but now it's explicit what operation each term is doing and what order to assemble the regression matrix. 
* In the future, the regular PDE library may become a superclass of this one to enable regular PDE-FIND of multiple trajectories and remove some of the hacks here.

## Future steps

1. Get JointSINDy merged
2. Find an alternate way to pass `_inputs_per_library`, enabling `GeneralizedLibrary`
3. Implement PDEFIND in a parent class between SINDy and this class, making `PDELibrary` merely a dataclass 
4. more symbolically-aware function libraries
5. some private methods shared by SINDy and WeakSINDy that SINDy-PI can invoke in order to correctly construct the feature library (rather than using `feature_library.transform()`, as now).

Remaining to do:
- [x] Rebuild SINDy-PI as a separate class that uses pysindy objects, rather than fitting awkwardly into other classes.
- [x] Find the missing -1 factor
- [x] Add docstrings
- [x] Implement partial simulate()/score() for ODEs
- [x] Add tests for `u_dot_wk` 
- [x] Add test for `eval_semiterm()`
- [x] Add test for local einsum fix
- [x] Check control
- [x] Add runnable example to the docstring.
- [x] ~Linting~ (Syntax for tests requires upgrading to 3.12, which requires upgrading pre-commits to conflicting versions.  This is a TODO: For afterwards).
- [x] Add LHS to `fit()` to allow steady-state solutions (e.g. $$f(x) = Du$$)
- [x] ~Remove or modify `reshape_samples_to_spatial_grid` now that we don't use it for weak (may still be needed for PDELibrary)~ Fixed PDE prediction shapes across `SINDy` and `WeakSINDy`.  Still probably need to remove this method, though.
- [x] Make `SemiTerm` a class, not just a tuple type alias
- [ ] Verify max derivative order for test function is correctly implemented
- [ ] Check that `_linear_weights` works for arbitrary test function

### CODE REVIEW

**Comment by Jacob-Stevens-Haas** [comment]
Because of the required linting updates for the syntax in this PR,, pre-commit --run-all finds issues outside this PR.   The right call is to modify the linting CI to pre-commit `--from-ref` and `--to-ref`, but for now I'll treat it like a pass, then after this PR implement a pre-commit update

File: pysindy/differentiation/finite_difference.py [hunk 1]
```diff
@@ -138,20 +146,24 @@ def _coefficients_boundary_forward(self, t):
...
+        pows = AxesArray(
+            np.arange(self.n_stencil_forward)[np.newaxis, :, np.newaxis, np.newaxis],
+            {"ax_time": 0, "ax_power": 1, "ax_offset": 2, "ax_coord": 3},
+        )
 
         if np.isscalar(t):
+            warn("Scalar t is deprecated here", DeprecationWarning, stacklevel=2)
             matrices = np.transpose(
-                (t * (self.stencil_inds - tinds)[:, np.newaxis, :]) ** pows
+                (t * (self.stencil_inds - tinds)[:, "power", ...]) ** pows
```
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/differentiation/finite_difference.py]: I don't believe this does raise since `t` isn't being indexed; `self.stencil_inds - tinds` is, which is always an AxesArray

File: asv_bench/benchmarks/defaults.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        self.model.fit([self.x], [self.t])
+        if not hasattr(self.model, "feature_names_"):
+            raise SkipNotImplemented
+        true_ode_align, est_ode_align = sindy_exp._utils.unionize_coeff_dicts(
+            self.model, self.true_eqns
+        )
+        metrics = sindy_exp.coeff_metrics(est_ode_align, true_ode_align)
+        return metrics["coeff_mae"]
+
+    def time_fit_predict(self):
```
Comment (yb6599) [inline asv_bench/benchmarks/defaults.py]: Do `time_fit_predict` and `peakmem_fit_predict` not return anything here? I'm guessing this is a WIP yet since none of the the `time` and `peakmem` functions return anything in this or `special.py`

File: pysindy/_sindypi.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            t = [np.arange(0, xi.shape[-2] * t, t) for xi in x]
+        t = [standardize_shape(ti) for ti in t]
+
+        if x_dot is not None:
+            x_dot = [standardize_shape(xdoti) for xdoti in x_dot]
+
+        _validate_inputs(x, t, x_dot, None)
+
+        if x_dot is None:
+            x_smooth, x_dot = zip(
```
Comment (yb6599) [inline pysindy/_sindypi.py]: NIT but should we also denote the output of `process_trajectories` in `SINDy.fit()` as `x_smooth` instead of `x` to keep naming consistent?

**Review by yb6599 — APPROVED** [review]
Hey Jake, the documentation and type annotations are thorough. Aside from some minor comments, LGTM.
